### PR TITLE
Common functions to support gMSA linux

### PR DESCRIPTION
**Pull request by saikiranakula-amzn**
### Summary
Moved functions to common files needed for supporting both windows and linux gMSA and refactored the s3 module in agent, to be more consistent with naming as per other aws modules in the code base

### Implementation details
Renamed existing s3client to s3managerclient
Added new s3client to support get object from s3
Moved credentialspec code to common go files

### Testing
Added unit tests
Note for external contributors:
`make test` and `make run-integ-tests` can run in a Linux development
environment like your laptop.  `go test -timeout=30s ./agent/...` and
`.\scripts\run-integ.tests.ps1` can run in a Windows development environment
like your laptop.  Please ensure unit and integration tests pass (on at least
one platform) before opening the pull request.
Once you open the pull request, there will be 14 automatic test checks on the bottom
of the pull request, please make sure they all pass before you merge it. You can
use `bot/test` label to rerun the automatic tests multiple times.
-->

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/s3/factory/factory.go [hunk 1]
```diff
@@ -62,6 +64,19 @@ func (*s3ClientCreator) NewS3ClientForBucket(bucket, region string,
 	return s3manager.NewDownloaderWithClient(s3.New(sessWithRegion)), nil
 }
 
+// NewS3Client returns a new S3 client
+func (*s3ClientCreator) NewS3Client(region string,
```
Comment (singholt) [inline agent/s3/factory/factory.go:68]: why not use the existing `NewS3ClientForBucket`?

Comment (saikiranakula-amzn) [reply agent/s3/factory/factory.go:68]: it only have the download file option. I don't need it, since I am not downloading anything on the instance

Comment (singholt) [reply agent/s3/factory/factory.go:68]: I see, can you please update/add a comment for this func and the above `NewS3ClientCreator()` to reflect this.

File: agent/s3/s3.go [hunk 2]
```diff
@@ -50,3 +52,24 @@ func ParseS3ARN(s3ARN string) (bucket string, key string, err error) {
...
+		Key:    aws.String(key),
+	}
+
+	result, err := client.GetObject(requestInput)
+	if err != nil {
+		return "", err
+	}
+
+	defer result.Body.Close()
+	resultBody, err := ioutil.ReadAll(result.Body)
```
Comment (singholt) [inline agent/s3/s3.go:68]: looks like ioutil is deprecated: https://pkg.go.dev/io/ioutil

you can use `io.ReadAll(...)` instead.

File: agent/s3/factory/factory.go [hunk 3]
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/aws/amazon-ecs-agent/agent/credentials"
 	"github.com/aws/amazon-ecs-agent/agent/httpclient"
 	s3client "github.com/aws/amazon-ecs-agent/agent/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
```
Comment (singholt) [inline agent/s3/factory/factory.go:22]: nit: i think this import needs to be clubbed in the block below, with other external imports

File: agent/s3/s3.go (hunk 2 above)
Comment (saikiranakula-amzn) [reply agent/s3/s3.go:68]: sure

File: agent/s3/factory/factory.go (hunk 1 above)
Comment (saikiranakula-amzn) [reply agent/s3/factory/factory.go:71]: added comments

File: agent/s3/s3_test.go [hunk 4]
```diff
@@ -78,3 +80,19 @@ func TestParseS3ARNInvalid(t *testing.T) {
...
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockGetObjectResponse := &s3sdk.GetObjectOutput{
+		Body: io.NopCloser(strings.NewReader(expectedValue)),
+	}
+	mockS3Client := mock_s3.NewMockS3Client(ctrl)
+	mockS3Client.EXPECT().GetObject(gomock.Any()).Return(mockGetObjectResponse, nil)
+
+	actualValue, err := GetObject("foo", "bar", mockS3Client)
```
Comment (singholt) [inline agent/s3/s3_test.go:95]: nit: you can use the existing `testBucket` and `testKey` constants.

File: agent/s3/s3_test.go [hunk 5]
```diff
@@ -78,3 +80,19 @@ func TestParseS3ARNInvalid(t *testing.T) {
 	_, _, err := ParseS3ARN("arn:aws:xxx:::xxx")
 	assert.Error(t, err)
 }
+
+func TestGetObject(t *testing.T) {
```
Comment (singholt) [inline agent/s3/s3_test.go:84]: can you add failure test case too?

Comment (saikiranakula-amzn) [reply agent/s3/s3_test.go:84]: added failure test

**Review by singholt — APPROVED** [review]

**Review by vsiddharth — APPROVED** [review]
🚀

**Review by sparrc — APPROVED** [review]

**Review by ubhattacharjya — APPROVED** [review]

**Pull request merged**